### ISSUE TITLE
Incrementally reading very large files

### ISSUE THREAD

**Issue by mlohry** [zeux/pugixml#166]
I use the VTK file formats in computational physics (http://www.vtk.org/VTK/img/file-formats.pdf) which are themselves valid XML files. They're very easy write in parallel with MPI-IO, but reading in serial can be an issue with size. For very large parallel cases I end up with XML files anywhere from 1-10+GB each, where most of the data is in several million leaf nodes differentiated by attributes. It looks like pugixml allocates storage for the entire file and fread's it in load_file_impl, which is obviously problematic if you don't have that amount of memory available to each process.

Is there any mechanism to load part of an XML file, ideally only loading the data from a node matching specific nodename+attr combination?

**Comment by tuananh** [comment]
siviucpp built a library based on pugixml for streaming parsing XML. i think it can handle your case

https://github.com/silviucpp/erlxml/blob/master/c_src/xmlstreamparser.h

**Comment by zeux** [comment]
There's no mechanism for reading files incrementally in pugixml. This is covered in #135 - TLDR is it might require a complete rewrite.

Here's the most memory efficient way to parse large files:

- Compile pugixml with PUGIXML_COMPACT define; this makes metadata pretty compact, 12b per node and 8b per attribute
- Open your file with mmap in copy-on-write mode (MAP_PRIVATE)
- Use load_buffer_inplace
- To avoid excessive page copies, use parse_minimal parsing flags (if your document has CDATA nodes, use parse_cdata)

mmap + copy-on-write does wonders for documents with relatively large values (not sure if VTK follows that); if the values are small then it'll still result in page copying for most of the pages due to string null-termination that pugixml does.

**Comment by mlohry** [comment]
So I'm clear by what you mean, with default loading pugixml will require at minimum 2X the file size in memory, but load_buffer_inplace will make that more like 1X the file size?

Understood that partial reads would require a complete rewrite. Might have to end up rolling my own on this =)

Thanks!

**Comment by zeux** [comment]
By default pugixml requires physical memory allocation for:

- All file data - this is equivalent to file size
- All metadata (node structure) - this is frequently larger than file size, especially in 64-bit mode

Using PUGIXML_COMPACT significantly reduces the second part; depending on the file structure it could still be as large as the entire file - as mentioned previously, 12b per node and 8b per attribute is pretty small but for XML files like `<n/>`+ each node is only 4 bytes in the source file...

The point of mmap+load_buffer_inplace is to try to reduce the first part by making it so that if there are large node/attribute values (>4K), they can consume no physical memory since copy-on-write mechanism will keep them backed by the file and kernel will load/unload these pages on demand, maintaining reasonable physical memory consumption. This is only beneficial if file has large values - if it doesn't then using load_file is just as good wrt memory consumption.

And yeah as @tuananh mentioned above, if the file has a logically "flat" structure, you can write a tokenizer that will identify block boundaries and push blocks of XML data into pugixml. This doesn't work for all documents, but might work for your usecase.

**Comment by zeux** [comment]
Closing as duplicate of #135 (there's no planned action beyond suggested workarounds here and in the linked issue).
